### ISSUE TITLE
Replace Binary Serializer with Newtonsoft Json Serialiser (if possible) or find work around

### ISSUE THREAD

**Issue by richard-churchman** [jube-home/jube#9]
In .Net 7 the BinarySerializer is deprecated.  The BinarySerializer is not used extensively,  except to save Neural Network model states.

The suggested replacement for BinarySerialization is Json,  however, it is not clear the extent to which this absolutely serializes an object,  including internal properties and the total state.

This ticket is to replace all transient and persistent serialized objects with a Json serialisation if possible, and figure out a workaround.

This refactoring is currently the main blocker in a path to .Net 7.

**Comment by richard-churchman** [comment]
Initial research seems positive and Newtonsoft would seem to achieve the same as something we created to do the same during model training, but has the advantage of being easier to persist to the database.  Mocked as below:

```
using System.Reflection;
using Newtonsoft.Json;
using Newtonsoft.Json.Serialization;
using TestingJsonSerialiser;

var test = new TestClass();

var settings = new JsonSerializerSettings() { ContractResolver = new TestContractResolver() };
var json = JsonConvert.SerializeObject(test, settings);

Console.WriteLine(json);

public class TestContractResolver : DefaultContractResolver
{
    protected override IList<JsonProperty> CreateProperties(Type type, MemberSerialization memberSerialization)
    {
        var props = type.GetProperties(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance)
            .Select(p => base.CreateProperty(p, memberSerialization))
            .Union(type.GetFields(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance)
                .Select(f => base.CreateProperty(f, memberSerialization)))
            .ToList();
        props.ForEach(p => { p.Writable = true; p.Readable = true; });
        return props;
    }
}
```

It appeared to handle Dictionaries, Lists and Byte Arrays for all public, private and instance fields.  Importantly, bytes go to Base64 and appear robust.

It should be pretty quick to create a helper class here,  ideal instantiated at the application start to ensure that the ContractResolver is in cache.

**Comment by richard-churchman** [comment]
Work underway.

**Comment by richard-churchman** [comment]
The only use of the Binary Serializer is the saving of Neural Network weights from Accord.net.  This project is archived and is not open for development.  Research has led me to conclude that json serialisation is still the best option however many of the classes do not have serialization attributes or total coverage of property seters.  Given the Accord.net project has gone into an archive state,  I think the best course of action is to include the required projects and code inside the Jube solution, and then modify the projects as required to maintain continuity.  I don't want to put in another Neural Network training library as this works pretty well, and the concepts have not seen radical change.

The next steps are to drop the Accord.net Nuget packages and instead reference the project locally in the Jube solution.  Then address each class and ensure that it better supports json serialisation.

This will also require the deprecation of the binary serialisation from Accord.net,  to the extent that it is not used outside of saving, which is unknown.

**Comment by richard-churchman** [comment]
All the Accord.net projects that are in use by Jube are now projects in the solution and build to .Net 6.  There remain problems in serialisation however as not all properties have setters and some private fields exist without exposing a property.  As I am already knee-deep Accord.Net code,  I would prefer to refactor to make sure that all private fields are available via a property Get \ Set methods.  I would also like to make sure that in the event that a field is in some wat derived it throws a Not Supported exception,  so that could be handled in Newtonsoft on deserialization.

**Comment by richard-churchman** [comment]
Through a fair amount of tinkering with setters, missing public properties and other minor refactoring,  the following json now serializes and deserialises via Newtonsoft json:

``` json
{
  "Layers": [
    {
      "$type": "Accord.Neuro.ActivationLayer, Accord.Neuro",
      "Output": [
        -1.0,
        -1.0,
        1.0,
        -1.0,
        -1.0
      ],
      "Neurons": [
        {
          "$type": "Accord.Neuro.ActivationNeuron, Accord.Neuro",
          "Output": -1.0,
          "Weights": [
            -0.0034046967048197985,
            0.049631912261247635,
            -0.14164109528064728,
            -0.21592922508716583,
            0.04197319969534874,
            -0.32003480195999146,
            -0.1483887881040573,
            -0.3135829567909241,
            0.09733650833368301,
            -0.006185198202729225,
            -0.012862536124885082,
            -0.3536251187324524,
            0.11490730196237564
          ],
          "Threshold": 0.31355854868888855,
          "InputsCount": 13,
          "RandGenerator": {
            "A": -0.5,
            "B": 0.5,
            "Mean": 0.0,
            "Mode": 0.0,
            "$type": "Accord.Statistics.Distributions.Univariate.UniformContinuousDistribution, Accord.Statistics",
            "Length": 1.0,
            "Median": 0.0,
            "Entropy": 0.0,
            "Maximum": 0.5,
            "Minimum": -0.5,
            "Support": {
              "Max": 0.5,
              "Min": -0.5,
              "Length": 1.0
            },
            "Variance": 0.08333333333333333,
            "Quartiles": {
              "Max": 0.25,
              "Min": -0.25,
              "Length": 0.5
            },
            "StandardDeviation": 0.28867513459481287
          },
          "ActivationFunction": {
            "$type": "Accord.Neuro.BipolarSigmoidFunction, Accord.Neuro",
            "Alpha": 2.0
          }
        },
        {
          "$type": "Accord.Neuro.ActivationNeuron, Accord.Neuro",
          "Output": -1.0,
          "Weights": [
            -0.20585231482982635,
            0.20159702003002167,
            -2.118711233139038,
            0.03948090225458145,
            -1.8429183959960938,
            0.19218377768993378,
            1.4026474952697754,
            -0.2773945927619934,
            1.383162260055542,
            0.050706177949905396,
            -0.18174158036708832,
            0.19223737716674805,
            -0.33502042293548584
          ],
          "Threshold": -2.901390790939331,
          "InputsCount": 13,
          "RandGenerator": {
            "A": -0.5,
            "B": 0.5,
            "Mean": 0.0,
            "Mode": 0.0,
            "$type": "Accord.Statistics.Distributions.Univariate.UniformContinuousDistribution, Accord.Statistics",
            "Length": 1.0,
            "Median": 0.0,
            "Entropy": 0.0,
            "Maximum": 0.5,
            "Minimum": -0.5,
            "Support": {
              "Max": 0.5,
              "Min": -0.5,
              "Length": 1.0
            },
            "Variance": 0.08333333333333333,
            "Quartiles": {
              "Max": 0.25,
              "Min": -0.25,
              "Length": 0.5
            },
            "StandardDeviation": 0.28867513459481287
          },
          "ActivationFunction": {
            "$type": "Accord.Neuro.BipolarSigmoidFunction, Accord.Neuro",
            "Alpha": 2.0
          }
        },
        {
          "$type": "Accord.Neuro.ActivationNeuron, Accord.Neuro",
          "Output": 1.0,
          "Weights": [
            5.113378524780273,
            -0.17282524704933167,
            4.834943771362305,
            0.09563280642032623,
            -11.161144256591797,
            0.173253133893013,
            -2.692711591720581,
            -0.17507949471473694,
            -3.3828330039978027,
            0.25240859389305115,
            0.4140300750732422,
            0.3256479501724243,
            0.33516019582748413
          ],
          "Threshold": -4.48732852935791,
          "InputsCount": 13,
          "RandGenerator": {
            "A": -0.5,
            "B": 0.5,
            "Mean": 0.0,
            "Mode": 0.0,
            "$type": "Accord.Statistics.Distributions.Univariate.UniformContinuousDistribution, Accord.Statistics",
            "Length": 1.0,
            "Median": 0.0,
            "Entropy": 0.0,
            "Maximum": 0.5,
            "Minimum": -0.5,
            "Support": {
              "Max": 0.5,
              "Min": -0.5,
              "Length": 1.0
            },
            "Variance": 0.08333333333333333,
            "Quartiles": {
              "Max": 0.25,
              "Min": -0.25,
              "Length": 0.5
            },
            "StandardDeviation": 0.28867513459481287
          },
          "ActivationFunction": {
            "$type": "Accord.Neuro.BipolarSigmoidFunction, Accord.Neuro",
            "Alpha": 2.0
          }
        },
        {
          "$type": "Accord.Neuro.ActivationNeuron, Accord.Neuro",
          "Output": -1.0,
          "Weights": [
            0.00807425007224083,
            -0.08084903657436371,
            -0.12551985681056976,
            0.31649544835090637,
            -0.021792719140648842,
            0.030917201191186905,
            0.41504669189453125,
            0.32764509320259094,
            0.07826090604066849,
            -0.2659287452697754,
            -0.009107157588005066,
            -0.34701427817344666,
            -0.061601921916007996
          ],
          "Threshold": 0.19849923253059387,
          "InputsCount": 13,
          "RandGenerator": {
            "A": -0.5,
            "B": 0.5,
            "Mean": 0.0,
            "Mode": 0.0,
            "$type": "Accord.Statistics.Distributions.Univariate.UniformContinuousDistribution, Accord.Statistics",
            "Length": 1.0,
            "Median": 0.0,
            "Entropy": 0.0,
            "Maximum": 0.5,
            "Minimum": -0.5,
            "Support": {
              "Max": 0.5,
              "Min": -0.5,
              "Length": 1.0
            },
            "Variance": 0.08333333333333333,
            "Quartiles": {
              "Max": 0.25,
              "Min": -0.25,
              "Length": 0.5
            },
            "StandardDeviation": 0.28867513459481287
          },
          "ActivationFunction": {
            "$type": "Accord.Neuro.BipolarSigmoidFunction, Accord.Neuro",
            "Alpha": 2.0
          }
        },
        {
          "$type": "Accord.Neuro.ActivationNeuron, Accord.Neuro",
          "Output": -1.0,
          "Weights": [
            0.007836229167878628,
            0.11239149421453476,
            -0.384717732667923,
            -0.2556014657020569,
            0.03957099840044975,
            0.21452608704566956,
            0.10455867648124695,
            -0.035511452704668045,
            0.25863876938819885,
            0.20386159420013428,
            -0.03328203037381172,
            0.26383692026138306,
            -0.11138642579317093
          ],
          "Threshold": -0.17877808213233948,
          "InputsCount": 13,
          "RandGenerator": {
            "A": -0.5,
            "B": 0.5,
            "Mean": 0.0,
            "Mode": 0.0,
            "$type": "Accord.Statistics.Distributions.Univariate.UniformContinuousDistribution, Accord.Statistics",
            "Length": 1.0,
            "Median": 0.0,
            "Entropy": 0.0,
            "Maximum": 0.5,
            "Minimum": -0.5,
            "Support": {
              "Max": 0.5,
              "Min": -0.5,
              "Length": 1.0
            },
            "Variance": 0.08333333333333333,
            "Quartiles": {
              "Max": 0.25,
              "Min": -0.25,
              "Length": 0.5
            },
            "StandardDeviation": 0.28867513459481287
          },
          "ActivationFunction": {
            "$type": "Accord.Neuro.BipolarSigmoidFunction, Accord.Neuro",
            "Alpha": 2.0
          }
        }
      ],
      "InputsCount": 13,
      "NeuronsCount": 5
    },
    {
      "$type": "Accord.Neuro.ActivationLayer, Accord.Neuro",
      "Output": [
        0.9970154768325774
      ],
      "Neurons": [
        {
          "$type": "Accord.Neuro.ActivationNeuron, Accord.Neuro",
          "Output": 0.9970154768325774,
          "Weights": [
            -2.1562821865081787,
            -0.019232111051678658,
            -0.029979640617966652,
            -1.1866471767425537,
            1.112209677696228
          ],
          "Threshold": 1.0330123901367188,
          "InputsCount": 5,
          "RandGenerator": {
            "A": -0.5,
            "B": 0.5,
            "Mean": 0.0,
            "Mode": 0.0,
            "$type": "Accord.Statistics.Distributions.Univariate.UniformContinuousDistribution, Accord.Statistics",
            "Length": 1.0,
            "Median": 0.0,
            "Entropy": 0.0,
            "Maximum": 0.5,
            "Minimum": -0.5,
            "Support": {
              "Max": 0.5,
              "Min": -0.5,
              "Length": 1.0
            },
            "Variance": 0.08333333333333333,
            "Quartiles": {
              "Max": 0.25,
              "Min": -0.25,
              "Length": 0.5
            },
            "StandardDeviation": 0.28867513459481287
          },
          "ActivationFunction": {
            "$type": "Accord.Neuro.BipolarSigmoidFunction, Accord.Neuro",
            "Alpha": 2.0
          }
        }
      ],
      "InputsCount": 5,
      "NeuronsCount": 1
    }
  ],
  "Output": [
    0.9970154768325774
  ],
  "InputsCount": 13,
  "LayersCount": 2,
  "NeuronsCount": [
    5,
    1
  ]
}
```

I have checked that for all of the objects that are used in the training of Neural Networks via Exhaustive their getters and setters are fully available on object properties.

There a new columns by the title json in the ExhaustiveSearchInstancePromotedTrialInstance table which is added by Migration class GitHubIssueBranch9.

**Comment by richard-churchman** [comment]
BinarySerialization deprecated from Accord.net code in Jube.

**Comment by richard-churchman** [comment]
Created a test for model accuracy based on the mock training data.  Currently either the model training,  serialisation or recall is flawed as only overfit or junk models are being created.

I have pushed draft commits to the branch which will be squashed before master.  The branch is a little way from going to master as it is not passing tests.  Much more time-consuming debugging needed.

There are some 300+ TODO items in the legacy Accord.Net code which causes alerts in the static analysis.  These will need changing to something less bothersome, while retaining the general idea and documentation.

**Comment by richard-churchman** [comment]
Given that there is more control over the Accrod.Net library I think we should include a lot more meta-data from the training run, such as input field names and certain statistics associated with them.  This will help pair up the values in recall which has long been a source of issue and uncertainty.

**Comment by richard-churchman** [comment]
This has now been merged to master.  In conclusion:

Included open-source Accord.net library code natively to the solution given the deprecated state of that project.  Accord.net will be supported inside Jube for own needs (such as upgrading to .Net 7+).

Accord.net libraries modified to compile to .Net 6.

Modified Accord.net libraries to better support Newtonsoft Json.net serialisation.

Included annotations in the Accord.net libraries to house important dependency statistics. Serialised json can be used independently of Jube to recall a Neural Network.

Removal of Binary Serializer totally as a precursor to the projects support of .Net 7+.

Extensive bug fixing in crucial functions of Neural Network training and recall.  While this part of the system remains in Alpha, it is substantially more fit for purpose.
